### PR TITLE
Update keras3 Softmax mask handling to be more numerically robust.

### PR DESCRIPTION
**Pull request by yangliyl-g**
The original impl of masking is essentially `(1.0 - mask) * -1e-9 + inputs`; this can be sensitive to numerical noise on `mask` (imagine if it is slightly off from either 1 or 0; then we would be adding a very large perturbation to the inputs).

Using comparison and where ops are much more numerically robust. Also for cases where we get floating point masks coming in, we add a binarization step to make it compatible w/ the `where` op.

### CODE REVIEW

**Review by hertschuh — APPROVED** [review]
Thanks for the improvement!

**Comment by hertschuh** [comment]
Oh, you'll need to reformat the code.

`ruff --config pyproject.toml format .` should do the job.

Otherwise the real way is:
https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md#generating-public-api-and-formatting-the-code

**Review by hertschuh — APPROVED** [review]

**Comment by yangliyl-g** [comment]
This PR was based on a deep and thorough investigation led by [Jaswanth Sreeram](https://www.linkedin.com/in/jaswanth-sreeram-b9117763) who prototyped the solution. the identification and analysis of where this issue came from is largely thanks to Jaswanth & other folks who contributed to the difficult investigation.
